src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -323,37 +366,141 @@ public long getOffsetSafeToCommitToKafka() {
     }
   }
 
-  /* Returns the offset Token persisted into snowflake */
+  /**
+   * Fetches the offset token from Snowflake.
+   *
+   * <p>It uses <a href="https://github.com/failsafe-lib/failsafe">Failsafe library </a> which
+   * implements retries, fallbacks and circuit breaker.
+   *
+   * <p>Here is how Failsafe is implemented.
+   *
+   * <p>Fetches offsetToken from Snowflake (Streaming API)
+   *
+   * <p>If it returns a valid offset number, that number is returned back to caller.
+   *
+   * <p>If {@link net.snowflake.ingest.utils.SFException} is thrown, we will retry for max 3 times.
+   * (Including the original try)
+   *
+   * <p>Upon reaching the limit of maxRetries, we will {@link Fallback} to opening a channel and
+   * fetching offsetToken again.
+   *
+   * <p>Please note, upon executing fallback, we might throw an exception too. However, in that case
+   * we will not retry.
+   *
+   * @return long offset token present in snowflake for this channel/partition.
+   */
   @VisibleForTesting
-  protected long fetchLatestCommittedOffsetFromSnowflake() {
-    long latestCommittedOffsetInSnowflake;
+  protected long fetchOffsetTokenWithRetry() {
+    final RetryPolicy<Long> offsetTokenRetryPolicy =
+        RetryPolicy.<Long>builder()
+            .handle(SFException.class)
+            .withDelay(DURATION_BETWEEN_GET_OFFSET_TOKEN_RETRY)
+            .withMaxAttempts(MAX_GET_OFFSET_TOKEN_RETRIES)
+            .onRetry(
+                event ->
+                    LOGGER.warn(
+                        "[RetryPolicy] retry for getLatestCommittedOffsetToken. Retry no:{}",
+                        event.getAttemptCount()))
+            .build();
+
+    // Read it from reverse order. Fetch offsetToken, apply retry policy and then fallback.
+    return Failsafe.with(getFallbackForGetOffsetTokenFailure())
+        .onFailure(
+            event ->
+                LOGGER.error(
+                    "[Failsafe] Failure to fetch offsetToken even after retry and fallback from"
+                        + " snowflake for channel:{}, elapsedTimeSeconds:{}",
+                    this.getChannelName(),
+                    event.getElapsedTime().get(SECONDS),
+                    event.getException()))
+        .compose(offsetTokenRetryPolicy)
+        .get(this::fetchLatestCommittedOffsetFromSnowflake);
+  }
+
+  /**
+   * Fallback function to execute when retries for fetching getOffsetToken for channel fails.
+   *
+   * <p>Fallback function is to re-open the channel and fetch the latestOffsetToken one more time
+   * after reopen is successful.
+   *
+   * <p>Fallback is only attempted on SFException
+   *
+   * @return the fallback function to execute when all retries from getOffsetToken have exhausted.
+   */
+  private Fallback<Long> getFallbackForGetOffsetTokenFailure() {
+    return Fallback.builder(
+            () -> {
+              openChannelForTable();
```

Comment:
This is incorrect actually, I will need to assign the this.channel with object returned from openChannelForTable. 
I will fix this.

Reply:
Fixed in revision.